### DISCUSSION TITLE
Building a page builder, any advice?

### DISCUSSION THREAD

**Discussion by torian257x** [statamic/cms#4462] [General]
I understand it is subjective / opinionated... but I explicitly welcome that discussion here

we are making the decision currently whether to wrap all components from our page builder into "section" components.

So the page builder is a simple replicator that can contain section components.
Those section components decide what background color, what padding, what width etc

inside those section components we add the actual components, say, sliders which are basically other replicators with fieldsets inside

text with image
lists

etc

does that sound like a reasonable approach?

I was facing issues when using
replicator -> bard -> replicator

https://github.com/statamic/cms/issues/4417

and am now a bit worried I will face a lot more similar bugs and issues like that. Is there maybe a different approach that avoids nesting things so deeply?

**Comment by jasonvarga** [comment]
It's a good approach. The most popular starter kit, [Peak](https://peak.studio1902.nl) uses it.

If there's a bug on the cms side, it'll get fixed eventually.

**Reply by torian257x** [reply]
the thing is peak uses less nesting. Peak uses
replicator (page builder) -> component.


I think the biggest nesting in peak is

replicator (page builder) -> bard -> component


whereas our approach would be
replicator (page builder) -> replicator (section) -> bard -> component

or
replicator (page builder) -> bard (section) -> replicator -> component
